### PR TITLE
Work around CUDA+Clang Thrust issue

### PR DESCRIPTION
**Pull request by masterleinad**
We are hitting https://github.com/NVIDIA/thrust/issues/1032
```
use of undeclared identifier 'va_printf'
          _CubLog("Invoking %s<<<%u, %d, %d, %lld>>>(), %llu items total, %d items per thread, %d SM occupancy, %d vshmem size, %d ptx_version \n",
          ^
/usr/local/cuda/include/cub/util_debug.cuh:151:42: note: expanded from macro '_CubLog'
            #define _CubLog(format, ...) va_printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, __VA_ARGS__);
```
 in https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/ArborX/detail/PR-782/5/pipeline.
The build uses `Clang 14.0.0` with `CUDA 11.0.3` on `Ubuntu 18.04`. Apparently, this is doesn't given us a recent enough `thrust` that includes https://github.com/thrust/cub/commit/a568ffa1fe061c20689934f119afd09beae820fd.

The workaround proposed here, is the same one we use in `ArborX`, see https://github.com/arborx/ArborX/blob/77e055395c9550cb42e5b90156ba659f060bd852/src/details/ArborX_DetailsSortUtils.hpp#L26-L65.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: algorithms/src/Kokkos_Sort.hpp [hunk 1]
```diff
@@ -66,8 +66,35 @@
...
+//
+// Defining _CubLog here allows us to avoid that code path, however disabling
+// some debugging diagnostics
+//
+// If _CubLog is already defined, we save it into KOKKKOS_CubLog_save, and
+// restore it at the end
+#ifdef _CubLog
+#define KOKKKOS_CubLog_save _CubLog
+#endif
+#define _CubLog
```
Comment (crtrott) [inline algorithms/src/Kokkos_Sort.hpp:86]: Won't this trigger "redefinition warning" so you might need to first undef after 84?

**Review by crtrott — CHANGES_REQUESTED** [review]

File: algorithms/src/Kokkos_Sort.hpp (hunk 1 above)
Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp:87]: 29048f9d5

**Comment by dalg24** [comment]
Please discuss whether the current coverage for Clang+CUDA builds is appropriate

**Comment by dalg24** [comment]
Is the current combination of Clang/CUDA versions appropriate?  Are they other versions we might want to add as nightly builds?

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
I think the current version is fine for Clang/CUDA combo.

**Review by aprokop — COMMENTED** [review]
Seems good to me. Our recent failure in ArborX was using Clang 14 with Cuda 11, so recent versions are still susceptible. The missing `#undef` is a nice catch.

File: algorithms/src/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -66,8 +66,36 @@
...
+#undef _CubLog
+#endif
+#define _CubLog
 #include <thrust/device_ptr.h>
 #include <thrust/sort.h>
+#undef _CubLog
+#ifdef KOKKKOS_CubLog_save
+#define _CubLog KOKKKOS_CubLog_save
+#undef KOKKKOS_CubLog_save
+#endif
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp:94]:
```suggestion
#pragma push_macro("_CubLog")
#ifdef _CubLog
#undef _CubLog
#endif
#define _CubLog
#include <thrust/device_ptr.h>
#include <thrust/sort.h>
#pragma pop_macro("_CubLog")
```

Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp:94]: Do we consider `#pragma push_macro` save to be used across all compilers?

**Comment by masterleinad** [comment]
> Is the current combination of Clang/CUDA versions appropriate? Are they other versions we might want to add as nightly builds?

This one is very close to our minimum requirements but I would consider adding the failing one to nightly builds.

File: algorithms/src/Kokkos_Sort.hpp (hunk 2 above)
Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:94]: Here it is guarded for Clang.
But to answer your question, I think it is fine with MSVC, GCC, and Clang (until proof of the contrary)

File: algorithms/src/Kokkos_Sort.hpp [hunk 3]
```diff
@@ -66,8 +66,32 @@
...
+//    error: use of undeclared identifier 'va_printf'
+// The exact combination of versions for Clang and Thrust (or CUDA) for this
+// failure was not investigated, however even very recent version combination
+// (Clang 10.0.0 and Cuda 10.0) demonstrated failure.
+//
+// Defining _CubLog here allows us to avoid that code path, however disabling
+// some debugging diagnostics
+//
+// If _CubLog is already defined, we save it into KOKKKOS_CubLog_save, and
+// restore it at the end
```
Comment (aprokop) [inline algorithms/src/Kokkos_Sort.hpp:82]: You need to change this comment as you use `push_macro` so there is no longer `_save` macro.

Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp]: d70f9376ff93847a2b3e3ae515377c76e686f620

**Review by dalg24 — APPROVED** [review]

**Pull request merged**